File: VoxCap_code/linux/compute_1overR_quad_mingyu.cpp
#include <string>
#include <algorithm>
#include <iostream>
#include <cmath>
#include <vector>
#include <malloc.h>
#include "compute_1overR_quad_mingyu.h"
#include "mex.h"
#include <iterator>

using namespace std;
template<class T>
int length(T& arr)
{
	return sizeof(arr) / sizeof(arr[0]);
}


template<class T>
double norm(T& arr)
{
	return sqrt(pow(arr[0], 2) + pow(arr[1], 2) + pow(arr[2], 2));
}

double compute_1overR_quad_mingyu(double src_cen[3], double obs_cen[3], double unit_normal_src[3], double unit_normal_obs[3], double smpl_wghts[], double smpl_pnts_temp[], int len_smpl)
{
	double y = 0.0;
	double int_res = 0.0;
	int plane_src = 0; int plane_obs = 0; int form_type = 0; int ii;
	//int len_smpl = length(smpl_wghts);

	double arr_smpl[4096 * 4];
	double* smpl_pnts = &arr_smpl[0];
	for (int ii = 0; ii < len_smpl * 4; ii++)
	{
		arr_smpl[ii] = *(smpl_pnts_temp + ii);
	}


	if (norm(unit_normal_src) - 1.0 > 1e-13)
	{
		y = norm(unit_normal_src);
		for (plane_src = 0; plane_src < 3; plane_src++)
		{
			unit_normal_src[plane_src] /= y;
		}
	}
	if (norm(unit_normal_obs) - 1.0 > 1e-13)
	{
		y = norm(unit_normal_obs);
		for (plane_obs = 0; plane_obs < 3; plane_obs++)
		{
			unit_normal_obs[plane_obs] /= y;
		}
	}

	if ((std::abs(std::abs(unit_normal_src[2]) - 1.0) < 1.0E-13) && (std::abs
	(unit_normal_src[1]) < 1.0E-13) && (std::abs(unit_normal_src[0]) < 1.0E-13))
	{
		plane_src = 1;
	}
	else if ((std::abs(std::abs(unit_normal_src[1]) - 1.0) < 1.0E-13) && (std::
		abs(unit_normal_src[2]) < 1.0E-13) && (std::abs(unit_normal_src[0]) < 1.0E-13))
	{
		plane_src = 2;
	}
	else if ((std::abs(std::abs(unit_normal_src[0]) - 1.0) < 1.0E-13) && (std::abs
	(unit_normal_src[1]) < 1.0E-13) && (std::abs(unit_normal_src[2]) < 1.0E-13))
	{
		plane_src = 3;
	}

	if ((std::abs(std::abs(unit_normal_obs[2]) - 1.0) < 1.0E-13) && (std::abs
	(unit_normal_obs[1]) < 1.0E-13) && (std::abs(unit_normal_obs[0]) < 1.0E-13))
	{
		plane_obs = 1;
	}
	else if ((std::abs(std::abs(unit_normal_obs[1]) - 1.0) < 1.0E-13) && (std::
		abs(unit_normal_obs[2]) < 1.0E-13) && (std::abs(unit_normal_obs[0]) < 1.0E-13))
	{
		plane_obs = 2;
	}
	else if ((std::abs(std::abs(unit_normal_obs[0]) - 1.0) < 1.0E-13) && (std::abs
	(unit_normal_obs[1]) < 1.0E-13) && (std::abs(unit_normal_obs[2]) < 1.0E-13))
	{
		plane_obs = 3;
	}

	if (plane_obs == plane_src)
	{
		form_type = 1;
	}
	else
	{
		form_type = 2;
	}

	if (form_type == 1)
	{
		if (plane_src == 1)
		{
			for (ii = 0; ii < len_smpl; ii++)
			{
				smpl_pnts[ii] += obs_cen[0];
				smpl_pnts[len_smpl + ii] += src_cen[0];
				smpl_pnts[len_smpl * 2 + ii] += obs_cen[1];
				smpl_pnts[len_smpl * 3 + ii] += src_cen[1];

				y = 1.0 / sqrt(pow(smpl_pnts[ii] - smpl_pnts[len_smpl + ii], 2) + pow(smpl_pnts[len_smpl * 2 + ii] - smpl_pnts[len_smpl * 3 + ii], 2) + pow(obs_cen[2] - src_cen[2], 2));
				y *= smpl_wghts[ii];
				int_res += y;

			}
		}
		else if (plane_src == 2)
		{
			for (ii = 0; ii < len_smpl; ii++)
			{
				smpl_pnts[ii] += obs_cen[0];
				smpl_pnts[len_smpl + ii] += src_cen[0];
				smpl_pnts[len_smpl * 2 + ii] += obs_cen[2];
				smpl_pnts[len_smpl * 3 + ii] += src_cen[2];
				y = 1.0 / sqrt(pow(smpl_pnts[ii] - smpl_pnts[len_smpl + ii], 2) + pow(smpl_pnts[len_smpl * 2 + ii] - smpl_pnts[len_smpl * 3 + ii], 2) + pow(obs_cen[1] - src_cen[1], 2));
				y *= smpl_wghts[ii];
				int_res += y;
			}
		}
		else if (plane_src == 3)
		{
			for (ii = 0; ii < len_smpl; ii++)
			{
				smpl_pnts[ii] += obs_cen[1];
				smpl_pnts[len_smpl + ii] += src_cen[1];
				smpl_pnts[len_smpl * 2 + ii] += obs_cen[2];
				smpl_pnts[len_smpl * 3 + ii] += src_cen[2];
				y = 1.0 / sqrt(pow(smpl_pnts[ii] - smpl_pnts[len_smpl + ii], 2) + pow(smpl_pnts[len_smpl * 2 + ii] - smpl_pnts[len_smpl * 3 + ii], 2) + pow(obs_cen[0] - src_cen[0], 2));
				y *= smpl_wghts[ii];
				int_res += y;
			}
		}
	}
	else if (form_type == 2)
	{
		if ((plane_src == 1) && (plane_obs == 2))
		{
			for (ii = 0; ii < len_smpl; ii++)
			{
				smpl_pnts[ii] += obs_cen[0];
				smpl_pnts[len_smpl + ii] += src_cen[0];
				smpl_pnts[len_smpl * 3 + ii] += src_cen[1];
				smpl_pnts[len_smpl * 2 + ii] += obs_cen[2];
				y = 1.0 / sqrt(pow(smpl_pnts[ii] - smpl_pnts[len_smpl + ii], 2) + pow(obs_cen[1] - smpl_pnts[len_smpl * 3 + ii], 2) + pow(smpl_pnts[len_smpl * 2 + ii] - src_cen[2], 2));
				y *= smpl_wghts[ii];
				int_res += y;
			}
		}
		else if ((plane_src == 1) && (plane_obs == 3))
		{
			for (ii = 0; ii < len_smpl; ii++)
			{
				smpl_pnts[ii] += src_cen[0];
				smpl_pnts[len_smpl * 2 + ii] += obs_cen[1];
				smpl_pnts[len_smpl + ii] += src_cen[1];
				smpl_pnts[len_smpl * 3 + ii] += obs_cen[2];
				y = 1.0 / sqrt(pow(obs_cen[0] - smpl_pnts[ii], 2) + pow(smpl_pnts[len_smpl * 2 + ii] - smpl_pnts[len_smpl + ii], 2) + pow(smpl_pnts[len_smpl * 3 + ii] - src_cen[2], 2));
				y *= smpl_wghts[ii];
				int_res += y;
			}
		}
		else if ((plane_src == 2) && (plane_obs == 3))
		{
			for (ii = 0; ii < len_smpl; ii++)
			{
				smpl_pnts[ii] += src_cen[0];
				smpl_pnts[len_smpl + ii] += obs_cen[1];
				smpl_pnts[len_smpl * 2 + ii] += obs_cen[2];
				smpl_pnts[len_smpl * 3 + ii] += src_cen[2];
				y = 1.0 / sqrt(pow(obs_cen[0] - smpl_pnts[ii], 2) + pow(smpl_pnts[len_smpl + ii] - src_cen[1], 2) + pow(smpl_pnts[len_smpl * 2 + ii] - smpl_pnts[len_smpl * 3 + ii], 2));
				y *= smpl_wghts[ii];
				int_res += y;
			}
		}
		else if ((plane_src == 2) && (plane_obs == 1))
		{
			for (ii = 0; ii < len_smpl; ii++)
			{
				smpl_pnts[ii] += obs_cen[0];
				smpl_pnts[len_smpl + ii] += src_cen[0];
				smpl_pnts[len_smpl * 2 + ii] += obs_cen[1];
				smpl_pnts[len_smpl * 3 + ii] += src_cen[2];
				y = 1.0 / sqrt(pow(smpl_pnts[ii] - smpl_pnts[len_smpl + ii], 2) + pow(smpl_pnts[len_smpl * 2 + ii] - src_cen[1], 2) + pow(obs_cen[2] - smpl_pnts[len_smpl * 3 + ii], 2));
				y *= smpl_wghts[ii];
				int_res += y;
			}
		}
		else if ((plane_src == 3) && (plane_obs == 1))
		{
			for (ii = 0; ii < len_smpl; ii++)
			{
				smpl_pnts[ii] += obs_cen[0];
				smpl_pnts[len_smpl + ii] += obs_cen[1];
				smpl_pnts[len_smpl * 2 + ii] += src_cen[1];
				smpl_pnts[len_smpl * 3 + ii] += src_cen[2];
				y = 1.0 / sqrt(pow(smpl_pnts[ii] - src_cen[0], 2) + pow(smpl_pnts[len_smpl + ii] - smpl_pnts[len_smpl * 2 + ii], 2) + pow(obs_cen[2] - smpl_pnts[len_smpl * 3 + ii], 2));
				y *= smpl_wghts[ii];
				int_res += y;
			}
		}
		else if ((plane_src == 3) && (plane_obs == 2))
		{
			for (ii = 0; ii < len_smpl; ii++)
			{
				smpl_pnts[ii] += obs_cen[0];
				smpl_pnts[len_smpl * 2 + ii] += src_cen[1];
				smpl_pnts[len_smpl + ii] += obs_cen[2];
				smpl_pnts[len_smpl * 3 + ii] += src_cen[2];
				y = 1.0 / sqrt(pow(smpl_pnts[ii] - src_cen[0], 2) + pow(obs_cen[1] - smpl_pnts[len_smpl * 2 + ii], 2) + pow(smpl_pnts[len_smpl + ii] - smpl_pnts[len_smpl * 3 + ii], 2));
				y *= smpl_wghts[ii];
				int_res += y;
			}
		}
	}
	return int_res;
}

File: VoxCap_code/one_over_R/compute_1overR_quad_mingyu.m
function [int_res] = compute_1overR_quad_mingyu(dx,src_cen,obs_cen,unit_normal_src,unit_normal_obs,smpl_pnts,smpl_wghts)
%---------------------------------------------------------
%generate 1/R integral matrix with Gauss-Legendre quadrature
%--------------------------------------------------------
%%INPUTS
%dx: resolution namely the dimension of every voxel
%num_smpl: the number of integral sample points
%src_cen: center coordinates of source panel
%obs_cen: center coordinates of observer panel
%unit_normal_src: unit normal vector of source panel
%unit_normal_obs: unit normal vector of observer panel
%smpl_pnts,smpl_wghts: gauss points and weights
%%OUTPUT
%int_res: the result of 1/R integral
%--------------------------------------------------------

eps2=1e-13;
if ((norm(unit_normal_src)-1) > eps2)
    unit_normal_src=unit_normal_src/(norm(unit_normal_src));
end
if ((norm(unit_normal_obs)-1) > eps2)
    unit_normal_obs=unit_normal_obs/(norm(unit_normal_obs));
end
plane_src=0;
if (abs(abs(unit_normal_src(3))-1) < eps2 && abs(unit_normal_src(2)) < eps2 && abs(unit_normal_src(1)) < eps2)
    % on xy plane
    %disp('Source panel is parallel to xy plane');
    plane_src=1;
elseif (abs(abs(unit_normal_src(2))-1) < eps2 && abs(unit_normal_src(3)) < eps2 && abs(unit_normal_src(1)) < eps2)
    % on xz plane
    %disp('Source panel is parallel to xz plane');
    plane_src=2;
elseif (abs(abs(unit_normal_src(1))-1) < eps2 && abs(unit_normal_src(2)) < eps2 && abs(unit_normal_src(3)) < eps2)
    % on yz plane
    %disp('Source panel is parallel to yz plane');
    plane_src=3;
end

if (plane_src==0)
    error('Error! Source panel should be parallel to xy, xz, or yz planes')
end
plane_obs=0;
if (abs(abs(unit_normal_obs(3))-1) < eps2 && abs(unit_normal_obs(2)) < eps2 && abs(unit_normal_obs(1)) < eps2)
    % on xy plane
    %disp('Observer panel is parallel to xy plane');
    plane_obs=1;
elseif (abs(abs(unit_normal_obs(2))-1) < eps2 && abs(unit_normal_obs(3)) < eps2 && abs(unit_normal_obs(1)) < eps2)
    % on xz plane
    %disp('Observer panel is parallel to xz plane');
    plane_obs=2;
elseif (abs(abs(unit_normal_obs(1))-1) < eps2 && abs(unit_normal_obs(2)) < eps2 && abs(unit_normal_obs(3)) < eps2)
    % on yz plane
    %disp('Observer panel is parallel to yz plane');
    plane_obs=3;
end

if (plane_obs==0)
    error('Error! Observer panel should be parallel to xy, xz, or yz planes')
end
if (plane_obs == plane_src)
    %disp('the panels are on the same plane')
    form_type=1;
else
    %disp('the planes of panels are orthogonal')
    form_type=2;
end

%     intervals=[obs_cen(1)-dx/2 obs_cen(1)+dx/2; src_cen(1)-dx/2 src_cen(1)+dx/2; ...
%         obs_cen(2)-dx/2 obs_cen(2)+dx/2; src_cen(2)-dx/2 src_cen(2)+dx/2; ...
%         obs_cen(3)-dx/2 obs_cen(3)+dx/2; src_cen(3)-dx/2 src_cen(3)+dx/2;];
%     int_shift(:) = (intervals(:,2)+intervals(:,1))*0.5;

if (form_type == 1)
    if (plane_src == 1)
        z1=src_cen(3);
        z2=obs_cen(3);
        smpl_pnts(:,1)=(smpl_pnts(:,1)+obs_cen(1));
        smpl_pnts(:,2)=(smpl_pnts(:,2)+src_cen(1));
        smpl_pnts(:,3)=(smpl_pnts(:,3)+obs_cen(2));
        smpl_pnts(:,4)=(smpl_pnts(:,4)+src_cen(2));
        Grfn=@(x,xpr,y,ypr) 1./(sqrt(((x-xpr).^2+(y-ypr).^2)+(z2-z1)^2));
%          Grfn=@(x,xpr,y,ypr) 1./(sqrt(((x-xpr).^2+(y-ypr).^2+(intervals(5,1)-intervals(6,1)).^2)));
%          temp=Grfn(smpl_pnts(:,1)+int_shift(1),smpl_pnts(:,2)+int_shift(2),smpl_pnts(:,3)+int_shift(3),smpl_pnts(:,4)+int_shift(4));
%          int_res=sum(temp.*smpl_wghts);
    elseif (plane_src == 2)
        y1=src_cen(2);
        y2=obs_cen(2);
        smpl_pnts(:,1)=(smpl_pnts(:,1)+obs_cen(1));
        smpl_pnts(:,2)=(smpl_pnts(:,2)+src_cen(1));
        smpl_pnts(:,3)=(smpl_pnts(:,3)+obs_cen(3));
        smpl_pnts(:,4)=(smpl_pnts(:,4)+src_cen(3));
        Grfn=@(x,xpr,z,zpr) 1./(sqrt(((x-xpr).^2+(z-zpr).^2)+(y2-y1)^2));
%             Grfn=@(x,xpr,z,zpr) 1./(sqrt(((x-xpr).^2+(z-zpr).^2+(intervals(4,1)-intervals(3,1)).^2)));
%             temp=Grfn(smpl_pnts(:,1)+int_shift(1),smpl_pnts(:,2)+int_shift(2),smpl_pnts(:,3)+int_shift(5),smpl_pnts(:,4)+int_shift(6));
%             int_res=sum(temp.*smpl_wghts);
    elseif (plane_src == 3)
        x1=src_cen(1);
        x2=obs_cen(1);
        smpl_pnts(:,1)=(smpl_pnts(:,1)+obs_cen(2));
        smpl_pnts(:,2)=(smpl_pnts(:,2)+src_cen(2));
        smpl_pnts(:,3)=(smpl_pnts(:,3)+obs_cen(3));
        smpl_pnts(:,4)=(smpl_pnts(:,4)+src_cen(3));
        Grfn=@(y,ypr,z,zpr) 1./(sqrt(((y-ypr).^2+(z-zpr).^2)+(x2-x1)^2));
%          Grfn=@(y,ypr,z,zpr) 1./(sqrt(((y-ypr).^2+(z-zpr).^2+(intervals(2,1)-intervals(1,1)).^2)));  
%          temp=Grfn(smpl_pnts(:,1)+int_shift(3),smpl_pnts(:,2)+int_shift(4),smpl_pnts(:,3)+int_shift(5),smpl_pnts(:,4)+int_shift(6));
%          int_res=sum(temp.*smpl_wghts);  
        
    end
    
elseif(form_type == 2)
    if (plane_src == 1 && plane_obs == 2)
        smpl_pnts(:,1)=(smpl_pnts(:,1)+obs_cen(1));
        smpl_pnts(:,2)=(smpl_pnts(:,2)+src_cen(1));
        smpl_pnts(:,4)=(smpl_pnts(:,4)+src_cen(2));
        smpl_pnts(:,3)=(smpl_pnts(:,3)+obs_cen(3));
%         Grfn=@(x,xpr,ypr,z) 1./(sqrt(((x-xpr).^2+(obs_cen(2)-ypr).^2)+(z-src_cen(3)).^2));
        Grfn=@(x,xpr,z,ypr) 1./(sqrt(((x-xpr).^2+(obs_cen(2)-ypr).^2)+(z-src_cen(3)).^2));
%         Grfn=@(x,xpr,z,ypr) 1./(sqrt(((x-xpr).^2+(obs_cen(2)-ypr).^2+(z-src_cen(3)).^2)));
%         temp=Grfn(smpl_pnts(:,1)+int_shift(1),smpl_pnts(:,2)+int_shift(2),smpl_pnts(:,3)+int_shift(5),smpl_pnts(:,4)+int_shift(4));
%         int_res=sum(temp.*smpl_wghts);    
    elseif (plane_src == 1 && plane_obs == 3)
        smpl_pnts(:,1)=(smpl_pnts(:,1)+src_cen(1));
        smpl_pnts(:,3)=(smpl_pnts(:,3)+obs_cen(2));
        smpl_pnts(:,2)=(smpl_pnts(:,2)+src_cen(2));
        smpl_pnts(:,4)=(smpl_pnts(:,4)+obs_cen(3));
%         Grfn=@(xpr,y,ypr,z) 1./(sqrt(((obs_cen(1)-xpr).^2+(y-ypr).^2)+(z-src_cen(3)).^2));
        Grfn=@(xpr,ypr,y,z) 1./(sqrt(((obs_cen(1)-xpr).^2+(y-ypr).^2)+(z-src_cen(3)).^2));
%         Grfn=@(xpr,ypr,y,z) 1./(sqrt(((obs_cen(1)-xpr).^2+(y-ypr).^2+(z-src_cen(3)).^2)));
%         temp=Grfn(smpl_pnts(:,1)+int_shift(2),smpl_pnts(:,2)+int_shift(4),smpl_pnts(:,3)+int_shift(3),smpl_pnts(:,4)+int_shift(5));
%         int_res=sum(temp.*smpl_wghts);
    
    elseif (plane_src == 2 && plane_obs == 3)
        smpl_pnts(:,1)=(smpl_pnts(:,1)+src_cen(1));
        smpl_pnts(:,2)=(smpl_pnts(:,2)+obs_cen(2));
        smpl_pnts(:,3)=(smpl_pnts(:,3)+obs_cen(3));
        smpl_pnts(:,4)=(smpl_pnts(:,4)+src_cen(3));
        Grfn=@(xpr,y,z,zpr) 1./(sqrt(((obs_cen(1)-xpr).^2+(y-src_cen(2)).^2)+(z-zpr).^2));
%         Grfn=@(xpr,y,z,zpr) 1./(sqrt(((obs_cen(1)-xpr).^2+(y-src_cen(2)).^2+(z-zpr).^2)));
%         temp=Grfn(smpl_pnts(:,1)+int_shift(2),smpl_pnts(:,2)+int_shift(3),smpl_pnts(:,3)+int_shift(5),smpl_pnts(:,4)+int_shift(6));
%         int_res=sum(temp.*smpl_wghts);
    
    elseif (plane_src == 2 && plane_obs == 1)
        smpl_pnts(:,1)=(smpl_pnts(:,1)+obs_cen(1));
        smpl_pnts(:,2)=(smpl_pnts(:,2)+src_cen(1));
        smpl_pnts(:,3)=(smpl_pnts(:,3)+obs_cen(2));
        smpl_pnts(:,4)=(smpl_pnts(:,4)+src_cen(3));
        Grfn=@(x,xpr,y,zpr) 1./(sqrt(((x-xpr).^2+(y-src_cen(2)).^2)+(obs_cen(3)-zpr).^2));
    
%         Grfn=@(x,xpr,y,zpr) 1./(sqrt(((x-xpr).^2+(y-src_cen(2)).^2+(obs_cen(3)-zpr).^2)));
%         temp=Grfn(smpl_pnts(:,1)+int_shift(1),smpl_pnts(:,2)+int_shift(2),smpl_pnts(:,3)+int_shift(3),smpl_pnts(:,4)+int_shift(6));
%         int_res=sum(temp.*smpl_wghts);
   
    elseif (plane_src == 3 && plane_obs == 1)
        smpl_pnts(:,1)=(smpl_pnts(:,1)+obs_cen(1));
        smpl_pnts(:,2)=(smpl_pnts(:,2)+obs_cen(2));
        smpl_pnts(:,3)=(smpl_pnts(:,3)+src_cen(2));
        smpl_pnts(:,4)=(smpl_pnts(:,4)+src_cen(3));
        Grfn=@(x,y,ypr,zpr) 1./(sqrt(((x-src_cen(1)).^2+(y-ypr).^2)+(obs_cen(3)-zpr).^2));
%           Grfn=@(x,y,ypr,zpr) 1./(sqrt(((x-src_cen(1)).^2+(y-ypr).^2+(obs_cen(3)-zpr).^2)));
%           temp=Grfn(smpl_pnts(:,1)+int_shift(1),smpl_pnts(:,2)+int_shift(3),smpl_pnts(:,3)+int_shift(4),smpl_pnts(:,4)+int_shift(6));
%           int_res=sum(temp.*smpl_wghts);   
    
    elseif (plane_src == 3 && plane_obs == 2)
        smpl_pnts(:,1)=(smpl_pnts(:,1)+obs_cen(1));
        smpl_pnts(:,3)=(smpl_pnts(:,3)+src_cen(2));
        smpl_pnts(:,2)=(smpl_pnts(:,2)+obs_cen(3));
        smpl_pnts(:,4)=(smpl_pnts(:,4)+src_cen(3));
%         Grfn=@(x,ypr,z,zpr) 1./(sqrt(((x-src_cen(1)).^2+(obs_cen(2)-ypr).^2)+(z-zpr).^2));
        Grfn=@(x,z,ypr,zpr) 1./(sqrt(((x-src_cen(1)).^2+(obs_cen(2)-ypr).^2)+(z-zpr).^2));
%           Grfn=@(x,z,ypr,zpr) 1./(sqrt(((x-src_cen(1)).^2+(obs_cen(2)-ypr).^2+(z-zpr).^2)));
%           temp=Grfn(smpl_pnts(:,1)+int_shift(1),smpl_pnts(:,2)+int_shift(5),smpl_pnts(:,3)+int_shift(4),smpl_pnts(:,4)+int_shift(6));
%           int_res=sum(temp.*smpl_wghts);         

    end
end
temp=Grfn(smpl_pnts(:,1),smpl_pnts(:,2),smpl_pnts(:,3),smpl_pnts(:,4));
% int_res=sum(smpl_wghts'*temp);
int_res=sum(temp.*smpl_wghts);

